ably/proto/protocol_message.go

```diff
@@ -6,9 +6,17 @@ import (
 	"time"
 )
 
+// TR3
 const (
-	FlagPresence Flag = iota + 1
-	FlagBacklog
+	FlagHasPresence       Flag = 0
```

Comment:
[TR3 says](https://docs.ably.io/client-lib-development-guide/features/#TR3):

> ProtocolMessage Flag enum has the following values, where a flag with value n is defined to be set if the bitwise AND of the flags field with 2ⁿ is nonzero

So those numbers are the _bit indexes_, starting at 0, whose corresponding bit being 1 means the flag is activated. They're not the _value of the bits_ themselves. For example, HAS_PRESENCE being 0 means that a Flag whose first bit is 1 will have the HAS_PRESENCE flag activated, but a Flag with value 0 doesn't.

Those Flag constants here aren't supposed to be bit those indexes, but bit _masks_, so that the `Has` method below works. So, they should be like this:

```go
FlagHasPresence Flag = 1 << 0
FlagHasBacklog Flag = 1 << 1
FlagResumed Flag = 1 << 2
FlagTransient Flag = 1 << 4
// etc.
```

I think naming doesn't help here. `Flag` type should be `Flags` or `FlagsField`; each individual bit in a `Flags`/`FlagsField` is a flag.

Reply:
You mean those numbers represent one or a group of bit indexes that should be present in the incoming number value with bit indexes at the same position.

Reply:
> You mean those numbers represent one or a group of bit indexes that should be present in the incoming number value with bit indexes at the same position.

Yes, exactly. It's like indexes in an array of booleans. Except the array of booleans is packed in a single integer of 64 bits (ie. 64 booleans).

> I feel I have come across this type of encoding where one number can represent multiple flags at the same time.

It's a common data structure called [bit field](https://en.wikipedia.org/wiki/Bit_field).

Reply:
Okay, so I feel the current solution should be OK. I can see that's how it's implemented in other clients too.